### PR TITLE
[Program: GCI] Find if user is currently in mentorship relation using GET /users

### LINKED ISSUE #232: Show if user is currently on a relation on the GET /users API for each user
## Description

As a developer consuming the backend api,
I need to know if the user is already occupied (in a current relation),
so that I can provide a better experience on the backend, showing the user is not currently available.

The idea is to return for each user entry from GET /users API response, a new value indicating if the user is available for new mentorship relations. For example, if the user is already enrolled in a current mentorship relation then the value should be `false`.

The value returned could had the following JSON key `is_available`.

To do this, the logic will be probably contained on the UserDAO class, and you won't have to add anything to the User API code.

_this issue is intended for the Google Code-in participants, it can be released after the program ends on January 23rd_

## Mocks

The GET /users should have an additional returning value, like this:
```
{
    (...),
    "is_available": true,
    (...)
}
```

### Update [Required]
- [ ] Return a new value on the GET /users API
### Enhancement to Update [Optional]
- [ ] Create at least one test for this new feature

## Definition of Done
- [ ] All of the required items are completed.
- [x] Approval by 1 mentor.

## Estimation
4 hours

### PR DESCRIPTION
**Pull request by daksha-aeer**
### Description
Find if user is currently in mentorship relation using GET /users endpoint.

#### Working
- Logic in /dao/user.py
```
user is available for relation if:
	not_in_relation AND (need_mentoring OR available_to_mentor)
```
- Model in models/user.py
- Documentation in swagger.json

#### Test cases
- test_list_users_api_resource_auth(), test_list_users_api_resource_verified_users()
  updated to work with 'is_available' parameter
- New test cases 'test_list_users_api_is_available_true()', 'test_list_users_api_is_available_false()'
  created

Fixes #232

### Type of Change:

- Code
- Quality Assurance

**Code/Quality Assurance Only**
- This change requires a documentation update (software upgrade on readme file)
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
1. Swagger
task43_1

2. Call GET /users endpoint
image

```sh
home@HP:~$ curl -X GET "http://127.0.0.1:5000/users" -H "accept: application/json" -H "Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpYXQiOjE1NzcxNzE5ODgsIm5iZiI6MTU3NzE3MTk4OCwianRpIjoiMGYzMzc1M2QtZDM2YS00YWE4LTljZDEtYzNlMjkzY2I5MzU4IiwiZXhwIjoxNTc3Nzc2Nzg4LCJpZGVudGl0eSI6NiwiZnJlc2giOmZhbHNlLCJ0eXBlIjoiYWNjZXNzIn0.XZso4DRncAMxKsQVkc6shYBkkvP9fwyCaEoMn5KFT8E"
[
    {
        "id": 1,
        "username": "gciuser",
        "name": "gciuser",
        "slack_username": null,
        "bio": null,
        "location": null,
        "occupation": null,
        "organization": null,
        "interests": null,
        "skills": null,
        "need_mentoring": true,
        "available_to_mentor": true,
        "is_available": true
    },
    {
        "id": 2,
        "username": "gciusertwo",
        "name": "gciusertwo",
        "slack_username": null,
        "bio": null,
        "location": null,
        "occupation": null,
        "organization": null,
        "interests": null,
        "skills": null,
        "need_mentoring": true,
        "available_to_mentor": true,
        "is_available": true
    },
    {
        "id": 3,
        "username": "gciuserthree",
        "name": "gciuserthree",
        "slack_username": null,
        "bio": null,
        "location": null,
        "occupation": null,
        "organization": null,
        "interests": null,
        "skills": null,
        "need_mentoring": true,
        "available_to_mentor": true,
        "is_available": true
    },
    {
        "id": 4,
        "username": "username",
        "name": "username",
        "slack_username": null,
        "bio": null,
        "location": null,
        "occupation": null,
        "organization": null,
        "interests": null,
        "skills": null,
        "need_mentoring": true,
        "available_to_mentor": true,
        "is_available": true
    },
    {
        "id": 5,
        "username": "typofixuser",
        "name": "typofixuser",
        "slack_username": null,
        "bio": null,
        "location": null,
        "occupation": null,
        "organization": null,
        "interests": null,
        "skills": null,
        "need_mentoring": true,
        "available_to_mentor": true,
        "is_available": true
    },
    {
        "id": 7,
        "username": "gciuserfive",
        "name": "gciuserfive",
        "slack_username": "string",
        "bio": "string",
        "location": "string",
        "occupation": "string",
        "organization": "string",
        "interests": "string",
        "skills": "string",
        "need_mentoring": false,
        "available_to_mentor": false,
        "is_available": false
    },
    {
        "id": 8,
        "username": "gciuserfivet",
        "name": "gciuserfivet",
        "slack_username": null,
        "bio": null,
        "location": null,
        "occupation": null,
        "organization": null,
        "interests": null,
        "skills": null,
        "need_mentoring": true,
        "available_to_mentor": true,
        "is_available": false
    },
    {
        "id": 9,
        "username": "gciusereighteen",
        "name": "gciusereighteen",
        "slack_username": null,
        "bio": null,
        "location": null,
        "occupation": null,
        "organization": null,
        "interests": null,
        "skills": null,
        "need_mentoring": true,
        "available_to_mentor": true,
        "is_available": true
    },
    {
        "id": 10,
        "username": "trialuser",
        "name": "trialuser",
        "slack_username": null,
        "bio": null,
        "location": null,
        "occupation": null,
        "organization": null,
        "interests": null,
        "skills": null,
        "need_mentoring": true,
        "available_to_mentor": true,
        "is_available": true
    },
    {
        "id": 11,
        "username": "tempname",
        "name": "tempname",
        "slack_username": null,
        "bio": null,
        "location": null,
        "occupation": null,
        "organization": null,
        "interests": null,
        "skills": null,
        "need_mentoring": true,
        "available_to_mentor": true,
        "is_available": true
    },
    {
        "id": 12,
        "username": "strijwoi",
        "name": "tgkrlvkp",
        "slack_username": null,
        "bio": null,
        "location": null,
        "occupation": null,
        "organization": null,
        "interests": null,
        "skills": null,
        "need_mentoring": true,
        "available_to_mentor": true,
        "is_available": true
    },
    {
        "id": 13,
        "username": "gciusertwenty",
        "name": "gciusertwenty",
        "slack_username": null,
        "bio": null,
        "location": null,
        "occupation": null,
        "organization": null,
        "interests": null,
        "skills": null,
        "need_mentoring": true,
        "available_to_mentor": true,
        "is_available": true
    }
]
```

3. Tests
```sh
python -m unittest discover tests
```
image

### Checklist:
**Delete irrelevant options.**

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] Any dependent changes have been merged
- [ ] Update Postman API at /docs folder
- [x] Update Swagger documentation and the exported file at /docs folder
- [ ] Update requirements.txt

**Code/Quality Assurance Only**
- [ ] My changes generate no new warnings 
- [ ] My PR currently breaks something (fix or feature that would cause existing functionality to not work as expected)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been published in downstream modules

### CODE REVIEW

File: app/api/dao/user.py [hunk 1]
```diff
@@ -153,7 +153,11 @@ def list_users(user_id, is_verified=None):
                     list_of_users += [user.json()]
         else:
             list_of_users = [user.json() for user in users_list]
-
+        # user is available for relation if: not in relation * (need_mentoring + available_to_mentor)
+        for user in list_of_users:
+            user['is_available'] = False
+            if user['membership_status'] == None and (user['need_mentoring'] or user['available_to_mentor']):
```
Comment (annabauza) [inline app/api/dao/user.py]: Is it one to one relation? One mentor can't have more than one mentees?

File: docs/swagger.json [hunk 2]
```diff
@@ -870,7 +870,8 @@
                 "organization",
                 "skills",
                 "slack_username",
-                "username"
+                "username",
+                "is_available"
```
Comment (annabauza) [inline docs/swagger.json:874]: She'll we modify swagger.json manually?

**Review by annabauza — CHANGES_REQUESTED** [review]
Can you please answer some questions I left you here?

File: app/api/dao/user.py (hunk 1 above)
Comment (daksha-aeer) [reply app/api/dao/user.py]: membership_status field was not giving proper value when member was in mentorship relation. I replaced this with a call to ```MentorshipRelationDAO.list_current_mentorship_relation```. Boolean logic remains the same

```py
            current_relation_response = MentorshipRelationDAO.list_current_mentorship_relation(user['id'])
            # find if not in relation
            if not isinstance(current_relation_response,MentorshipRelationModel):
                if (user['need_mentoring'] or user['available_to_mentor']):
                    user['is_available'] = True
```

This becomes a one to one relation. ```if not isinstance(current_relation_response,MentorshipRelationModel)``` will return false if user is already in a current relation. Morever if user is not in a current relation he must also be available to be mentored / to be a mentor.

File: docs/swagger.json (hunk 2 above)
Comment (daksha-aeer) [reply docs/swagger.json:874]: I had done this in a previous task. The PR template also has a field ```Update Swagger documentation and the exported file at /docs folder```. Is there any way to automatically generate it?

**Review by annabauza — APPROVED** [review]

**Comment by annabauza** [comment]
cosing due to merge #360

**Pull request closed without merging**